File: algorithms/swift-algorithm-club/03_Two_Sum_Input-Array-Is-Sorted.go
/**
 * Two Sum Input Array Is Sorted
 */
package main

import (
	"fmt"
)

func main() {
	fmt.Println(twoSumSorted([]int{1, 2, 3, 4, 5}, 7))
	fmt.Println(twoSumSorted([]int{1, 3, 7, 9, 11}, 13))
	fmt.Println(twoSumSorted([]int{1, 7, 8, 9}, 16))
}

func twoSumSorted(nums []int, target int) []int {
	if len(nums) < 2 {
		return nil
	}

	for i, j := 0, len(nums)-1; i < j; {
		println("~~~", i, j)
		sum := nums[i] + nums[j]

		switch {
		case sum == target:
			return []int{i, j}
		case sum < target:
			i++
		default:
			j--
		}
	}

	return nil
}

File: algorithms/swift-algorithm-club/03_Two_Sum_Input_Array_Is_Sorted.rs
/**
 * Two Sum Input Array Is Sorted
 */

fn main() {
    println!("{:?}", two_sum_sorted(&vec![1, 2, 3, 4, 5], 7));
    println!("{:?}", two_sum_sorted(&vec![1, 3, 7, 9, 11], 13));
    println!("{:?}", two_sum_sorted(&vec![1, 7, 8, 9], 16));
}

fn two_sum_sorted(nums: &Vec<usize>, target: usize) -> Result<Vec<usize>, String> {
    if nums.len() < 2 {
        return Err("not found".to_string());
    }

    let (mut i, mut j) = (0, nums.len() - 1);

    while i < j {
        println!("~~~ {}, {}", i, j);
        let sum = nums[i] + nums[j];

        match sum {
            _ if sum == target => {
                return Ok(vec![i, j]);
            }
            _ if (sum < target) => i += 1,
            _ => j -= 1,
        }
    }

    return Err("not found".to_string());
}
